Wikipedia/Code/AutoLayoutSafeMultiLineButton.swift

```diff
@@ -0,0 +1,49 @@
+
+// See multi-line button thread: https://stackoverflow.com/q/23845982/135557
+@IBDesignable
+class AutoLayoutSafeMultiLineButton: UIButton {
+    func multiLineSafeSetup () {
+        titleLabel?.numberOfLines = 0
+        titleLabel?.lineBreakMode = .byWordWrapping
+        titleLabel?.adjustsFontSizeToFitWidth = false
+        setContentHuggingPriority(.defaultLow + 1, for: .vertical)
+        setContentHuggingPriority(.defaultLow + 1, for: .horizontal)
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        super.init(coder: aDecoder)
+        multiLineSafeSetup()
+    }
+    
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        multiLineSafeSetup()
+    }
+    
+    override var intrinsicContentSize: CGSize {
+        guard let titleLabel = titleLabel else {
+            return super.intrinsicContentSize
+        }
+        return sizeByAddingInsets(to: titleLabel.intrinsicContentSize)
+    }
+    
+    override func sizeThatFits(_ size: CGSize) -> CGSize {
+        guard let titleLabel = titleLabel else {
+            return super.sizeThatFits(size)
+        }
+        return sizeByAddingInsets(to: titleLabel.sizeThatFits(size))
+    }
+    
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        guard let titleLabel = titleLabel else { return }
+        titleLabel.preferredMaxLayoutWidth = titleLabel.frame.width
```

Comment:
Setting `preferredMaxLayoutWidth` likely affects the layout, so it should be done before `layoutSubviews` - I'm not sure these changes would be applied if there wasn't another layout pass after it was updated

This could be updated on bounds change by overriding bounds and updating it in the `didSet`. It would be dictated by the bounds width minus the `contentEdgeInsets`, `titleEdgeInsets`, the image width & image edge insets if an image is set (using `image(for:` not `imageView.image`). This way it's updated before the layout pass and doesn't rely on the `titleLabel.frame`

Reply:
oh cool will try!

Reply:
hmm so far no luck with this... but if i move the `preferredMaxLayoutWidth` before `super.layoutSubviews()` it works fine... would that be less likely to affect layout?

Reply:
I'd be leery of updating anything that could trigger a `setNeedsLayout` inside of `layoutSubviews` - potentially triggers an infinite layout loop

Reply:
I added a recursion guard and `Changes welcome` label ;)